new-components/reporters/pdf/internal/reporter/reporter.go

```
@@ -0,0 +1,204 @@
+	if err != nil {
+		return "", nil, fmt.Errorf("could not get current working directory: %w", err)
+	}
+
+	reportHTMLPath := filepath.Join(currentPath, "report.html")
+	//#nosec: G304
+	f, err := os.OpenFile(reportHTMLPath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0o600) //#nosec: G304
+	if err != nil {
+		return "", nil, fmt.Errorf("could not open report.html: %w", err)
+	}
+	if err = tmpl.Execute(f, data); err != nil {
+		return "", nil, fmt.Errorf("could not apply data to template: %w", err)
+	}
+	// close the file after writing it
+	defer func(f *os.File) {
+		err := f.Close()
+		if err != nil {
+			slog.Error("could not close file", slog.String("err", err.Error()))
+		}
+	}(f)
+
+	//todo: add instance id to name of file
+	reportPDFPath := filepath.Join(currentPath, "report.pdf")
+	reportPage := fmt.Sprintf("file:///%s", reportHTMLPath)
+	pdfBytes, err := pw.GetPDFOfPage(reportPage, reportPDFPath)
+	if err != nil {
+		return "", nil, fmt.Errorf("could not generate pdf from page %s, err: %w", reportPage, err)
+
+	}
+
+	// delete the intermediate HTML file
+	if err := os.Remove(reportHTMLPath); err != nil {
+		slog.Error("could not delete report.html", slog.String("err", err.Error()))
+	}
+	return reportPDFPath, pdfBytes, err
+}
+
+// FormatTime is a template function for the PDF, that converts a timestamp to a human-readable format
+func FormatTime(timestamp *int64) string {
+	if timestamp == nil {
+		return ""
+	}
+
+	// Convert the int64 value to a time.Time
+	parsedTime := time.Unix(*timestamp, 0)
+
+	// Format the time using a predefined layout
+	return parsedTime.Format(time.DateTime)
+}
+
+// uploadToS3 uploads the PDF to AWS
+func (p PdfReporter) uploadToS3(resultFilename string, pdfBytes []byte) error {
+	if p.conf.Bucket == "" {
+		slog.Error("bucket is empty, you need to provide a bucket name")
+	}
+
+	if p.conf.Region == "" {
+		slog.Error("region is empty, you need to provide a region name")
+	}
+	client, err := s3client.NewClient(p.conf.Region)
```

I'd initialise this too in the main so we can fail early in case the feature is enabled :)